filip/utils/cleanup.py

```diff
@@ -0,0 +1,157 @@
+"""
+Functions to clean up a tenant within a fiware based platform.
+
+created Oct 08, 2021
+
+@author Thomas Storek
+"""
+import functools
+from requests import RequestException
+from filip.models import FiwareHeader
+from filip.clients.ngsi_v2 import ContextBrokerClient, IoTAClient, QuantumLeapClient
+
+
+def clear_cb(url: str, fiware_header: FiwareHeader):
+    """
+    Function deletes all entities, registrations and subscriptions for a
+    given fiware header
+
+    Note:
+        Always clear the devices first because the IoT-Agent will otherwise
+        through errors if it cannot find its registration anymore.
+
+    Args:
+        url: Url of the context broker service
+        fiware_header: header of the tenant
+
+    Returns:
+        None
+    """
+    # create client
+    client = ContextBrokerClient(url=url, fiware_header=fiware_header)
+
+    # clear entities
+    entities = client.get_entity_list()
+    if entities:
+        client.update(entities=entities, action_type='delete')
+
+    # clear subscriptions
+    for sub in client.get_subscription_list():
+        client.delete_subscription(subscription_id=sub.id)
+
+    # clear registrations
+    for reg in client.get_registration_list():
+        client.delete_registration(registration_id=reg.id)
+
+
+def clear_iota(url: str, fiware_header: FiwareHeader):
+    """
+    Function deletes all device groups and devices for a
+    given fiware header
+
+
+
+    Args:
+        url: Url of the context broker service
+        fiware_header: header of the tenant
+
+    Returns:
+        None
+    """
+    # create client
+    client = IoTAClient(url=url, fiware_header=fiware_header)
+
+    # clear groups
+    for group in client.get_group_list():
+        client.delete_group(resource=group.resource,
+                            apikey=group.apikey)
+
```

Comment:
Is a device always part of a group?

Reply:
Nevermind, i was confused by the comment "#clear registrations", what should probably be "#clear devices". Therefore I thought we only clear the devices here indirectly over their groups.